### PR TITLE
[REVIEW] Move utilities to legacy

### LINKED ISSUE #3361: [FEA] Move contents of `src/utilities/` to `legacy/`
**Is your feature request related to a problem? Please describe.**

The majority of the contents of [`src/utilities` ](https://github.com/rapidsai/cudf/tree/branch-0.11/cpp/src/utilities)is legacy code that should no longer be used.

**Describe the solution you'd like**
 The files in `src/utilities` should all be moved into `src/utilities/legacy` and associated file paths updated. 

If there are any utilities worth saving, they should be moved to an appropriate header in `cudf/include/detail/utilities`.

### PR DESCRIPTION
**Pull request by rgsl888prabhu**
close #3361

### CODE REVIEW

**Comment by rgsl888prabhu** [comment]
> I think `int_fastdiv.h` (used by hashing) and `trie.h` (used by CSV reader) are independent of cudf types and should not be moved to legacy.
> 
> However, they should probably be moved to the modules that use them if they are only used in one place (e.g. out of `include` into appropriate location in `src`.)

Along with those, there are atomic_device operations have been defined which are being used, so I was planning to take care of those as well in a follow-up PR. Plugging along with all these changes felt as if it will be too much of a hassle for reviewer.

**Comment by rgsl888prabhu** [comment]
Taken care of copying essentials from `legacy` to `detail/utilities` and made necessary changes required.

File: cpp/src/iterator/transform_unary_functions.cuh [hunk 1]
```diff
@@ -28,7 +28,7 @@
 #define CUDF_TRANSFORM_UNARY_FUNCTIONS_CUH
 
 #include <cudf/cudf.h>
-#include <utilities/cudf_utils.h>       // need for CUDA_HOST_DEVICE_CALLABLE
+//#include <utilities/legacy/cudf_utils.h>       // need for CUDA_HOST_DEVICE_CALLABLE
```
Comment (harrism) [inline cpp/src/iterator/transform_unary_functions.cuh]: Delete if unneeded.

File: cpp/src/replace/replace.cu [hunk 2]
```diff
@@ -22,13 +22,10 @@
 #include <rmm/rmm.h>
 #include <cudf/types.hpp>
 #include <cudf/utilities/error.hpp>
-#include <cudf/utilities/legacy/type_dispatcher.hpp>
+//#include <cudf/utilities/legacy/type_dispatcher.hpp>
```
Comment (harrism) [inline cpp/src/replace/replace.cu]: Remove if unneeded.

**Comment by rgsl888prabhu** [comment]
> Would you mind updating java/src/main/native/include/jni_utils.hpp to fix the include of column_utils.hpp? Otherwise we'll have to followup this PR to fix the Java build.

@jlowe Taken care of the java file.

**Comment by harrism** [comment]
@jlowe please approve.

**Pull request merged**